From: Steve Dickson <SteveD@redhat.com>
Date: Tue, 2 Feb 2010 12:17:58 -0500
Subject: [nfs] NFSD: Create PF_INET6 listener in write_ports
Message-id: <4B6817F6.7060006@RedHat.com>
Patchwork-id: 23097
O-Subject: [RHEL6] [PATCH 10/10] NFSD: Create PF_INET6 listener in write_ports
Bugzilla: 560785
RH-Acked-by: Peter Staubach <staubach@redhat.com>
RH-Acked-by: Jeff Layton <jlayton@redhat.com>

commit 37498292aa97658a5d0a9bb84699ce8c1016bb74
Author: Chuck Lever <chuck.lever@oracle.com>
Date:   Tue Jan 26 14:04:22 2010 -0500

    NFSD: Create PF_INET6 listener in write_ports

    Try to create a PF_INET6 listener for NFSD, if IPv6 is enabled in the
    kernel.

    Make sure nfsd_serv's reference count is decreased if
    __write_ports_addxprt() failed to create a listener.  See
    __write_ports_addfd().

    Our current plan is to rely on rpc.nfsd to create appropriate IPv6
    listeners when server-side NFS/IPv6 support is desired.  Legacy
    behavior, via the write_threads or write_svc kernel APIs, will remain
    the same -- only IPv4 listeners are created.

    Signed-off-by: Chuck Lever <chuck.lever@oracle.com>
    [bfields@citi.umich.edu: Move error-handling code to end]
    Signed-off-by: J. Bruce Fields <bfields@citi.umich.edu>

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/fs/nfsd/nfsctl.c b/fs/nfsd/nfsctl.c
index f43ecd6..0f0e77f 100644
--- a/fs/nfsd/nfsctl.c
+++ b/fs/nfsd/nfsctl.c
@@ -988,6 +988,7 @@ static ssize_t __write_ports_delfd(char *buf)
 static ssize_t __write_ports_addxprt(char *buf)
 {
 	char transport[16];
+	struct svc_xprt *xprt;
 	int port, err;
 
 	if (sscanf(buf, "%15s %4u", transport, &port) != 2)
@@ -1003,8 +1004,23 @@ static ssize_t __write_ports_addxprt(char *buf)
 	err = svc_create_xprt(nfsd_serv, transport,
 				PF_INET, port, SVC_SOCK_ANONYMOUS);
 	if (err < 0)
-		return err;
+		goto out_err;
+
+	err = svc_create_xprt(nfsd_serv, transport,
+				PF_INET6, port, SVC_SOCK_ANONYMOUS);
+	if (err < 0 && err != -EAFNOSUPPORT)
+		goto out_close;
 	return 0;
+out_close:
+	xprt = svc_find_xprt(nfsd_serv, transport, PF_INET, port);
+	if (xprt != NULL) {
+		svc_close_xprt(xprt);
+		svc_xprt_put(xprt);
+	}
+out_err:
+	/* Decrease the count, but don't shut down the service */
+	nfsd_serv->sv_nrthreads--;
+	return err;
 }
 
 /*
